### ISSUE TITLE
Reader: "Invalid aria prop `aria-description` on <button> tag" warning in console

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero#4310]
I'm not sure if this is because `aria-description` is [still in draft](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-description) or because we're on an outdated version of React.

Is `aria-description` actually functional in Fx115? If so, we should see if upgrading React avoids this.

**Comment by abaevbog** [comment]
Yeah, we had this discussed in https://github.com/zotero/reader/pull/127#issuecomment-2163588247

Essentially, this must be just this react bug https://github.com/facebook/react/issues/21035 where it throws a warning that `aria-description` is not a valid aria property even though it is and it works very much as it should. For example, if you focus the any of the tags in the sidebar of the reader with voiceover, it does say "Filter annotations by this tag", which is the `aria-description` bit.

We are currently on react 17 and upgrading react to react 18 should fix the issue.

Alternatively, @mrtcode [mentioned](https://github.com/zotero/reader/pull/127#issuecomment-2167285094) the possibility of having the warning just suppressed?

**Comment by mrtcode** [comment]
I tried to suppress this error message but it was harder than I thought. I think we should just upgrade React.

**Comment by abaevbog** [comment]
Gotcha! I'll look into it. I suppose we would have to switch to react 18 at some point.

**Comment by abaevbog** [comment]
So I am working through [migrating to React 18](https://react.dev/blog/2022/03/08/react-18-upgrade-guide), and I wanted to check in on a couple of things to make sure I don't veer off the track.

1. Despite my initial idea, the react version in `package.json` of the reader did not affect that warning - it is the react version in the actual main repo. I guess given this is the react that is exposed in `resource://zotero/react.js` it makes sense...
2. The `root.render()` does not have a callback anymore, so [awaiting for the render to complete](https://github.com/zotero/zotero/blob/2849913d1b8aed69ef1a2361c5a41a68eea9279b/chrome/content/zotero/collectionTree.jsx#L45) as in, say, `CollectionTree.init()` is no longer possible. In the [discussion on it](https://github.com/reactwg/react-18/discussions/5#discussion-3385584) that I ran into, suggested alternatives were to pass a callback as a ref or to just skip a tick with `setTimeout`. To me, the best approximation for our use case seemed to await for a [promise](https://github.com/abaevbog/zotero/blob/ffd55104dfdb2509cb2f713335bedc2d4fb1b29c/chrome/content/zotero/collectionTree.jsx#L50) created at the top of `CollectionTree.init()` that would be [resolved](https://github.com/abaevbog/zotero/blob/ffd55104dfdb2509cb2f713335bedc2d4fb1b29c/chrome/content/zotero/collectionTree.jsx#L142) in `componentDidMount()`. Does that make sense? Alternatively, I did try to just add `await Zotero.Promise.delay()` before returning `CollectionTree.init` and it also seems to work fine - are there any reasons to not simply do that?
3. Currently, `"react-dom-server"` [exposes](https://github.com/zotero/zotero/blob/main/resource/react-dom-server.js) a specific file in `node_modules` for export. In react 18, the `renderToStaticMarkup` function is not in that specific file anymore, so [this export in icons.jsx](https://github.com/zotero/zotero/blob/2849913d1b8aed69ef1a2361c5a41a68eea9279b/chrome/content/zotero/components/icons.jsx#L4) breaks. Is [exposing the file](https://github.com/abaevbog/zotero/commit/ffd55104dfdb2509cb2f713335bedc2d4fb1b29c#diff-e066ee56b56181307d972eeea4a9f0ff6d1dd7702cd59c8565d807736355214c) with an [updated](https://github.com/abaevbog/zotero/commit/ffd55104dfdb2509cb2f713335bedc2d4fb1b29c#diff-58f25f92bea511387c786391d119feb1deb587618d8fb7dee43328ee2caa0cad) require statement the right way around this? Do I need to add any build-specific steps here?

For reference, this is my one [commit](https://github.com/abaevbog/zotero/commit/ffd55104dfdb2509cb2f713335bedc2d4fb1b29c#diff-e066ee56b56181307d972eeea4a9f0ff6d1dd7702cd59c8565d807736355214c) with updated `collectionTree` loading and that updated react export setup. I didn't feel like this is close enough to start even a draft PR but if this makes sense, it should be quite simple to extrapolate this approach to everything else.

**Comment by dstillman** [comment]
I'll defer to @tnajdek on React questions, but I'd say we should definitely wait for proper mounting and not just use a delay.

**Comment by dstillman** [comment]
(And we'd want to update the reader React version too after this, but I think that would only be relevant for the web library (if even that — maybe just for standalone/dev installs).)

**Comment by tnajdek** [comment]
> the react version in package.json of the reader did not affect that warning

For "zotero" build Reader configures React as external:
https://github.com/zotero/reader/blob/8836db602c0d4fe26e56266d6411bc7a77b94c96/webpack.config.js#L113C1-L121

For "web" it bundles "React" which sadly means users have to download it twice. We should probably address that in future - web library has been on React 18 for quite a while now.

> The root.render() does not have a callback anymore...

I didn't test this but passing callback as a ref should work? Something along this lines:

```
const root = createRoot(domEl);
await new Promise(resolve => 
    root.render(<CollectionTree ref={c => ref = c; resolve(); } {...opts } />);
);
		
```

> Currently, "react-dom-server" [exposes](https://github.com/zotero/zotero/blob/main/resource/react-dom-server.js) a specific file in node_modules for export...

One solution would be to stop using `getDOMElement` entirely (there is like 5 places where we still use that) and replace it with `getCSSIcon` instead.


Happy to help with any of the above, just let me know!

**Comment by abaevbog** [comment]
@tnajdek thank you, that clarifies things quite a bit! 

> I didn't test this but passing callback as a ref should work? Something along this lines:

Yes, this approach is certainly cleaner and it does seem to work!

> One solution would be to stop using getDOMElement entirely (there is like 5 places where we still use that) and replace it with getCSSIcon instead.

It does makes sense - `getDOMElement` was only used for a few remaining `png` icons that were not swapped for `svg`s yet. I think we can remove most of the logic related to it and a small special case for `getCSSIcon` to handle those particular icons.

I think I am on the right track! I took care of all the deprecation warnings in the main repo (reader and note-editor need to be tweaked to not show those warning too). I am seeing some test breakage in that PR, so I'll just need clean that up
